### PR TITLE
Dynamic chart release / Nas 108690

### PR DESCRIPTION
**Pull request by AlexKarpov98**
WIP: Working on dynamic chart release

### CODE REVIEW

**Review by sonicaj — COMMENTED** [review]
Good Job, i have following concerns:
1) Can you please clarify why we have `asterisk` at the end of some fields ?
Screenshot 2021-01-18 at 10 41 23 PM
2) If type of a variable is `hostPath` only then do we show the path selector
```
        - variable: datadir
          label: "Nextcloud data directory"
          schema:
            type: path
            default: "/var/www/html/data"
            required: true
```
So for above, we shouldn't show file selector widget.
3) We are not validating `show_if` conditionals, for example consider the following example
```
  - variable: nextcloudDataHostPathEnabled
    label: "Configure Host Path for Nextcloud data"
    group: "Storage"
    schema:
      type: boolean
      default: false
      show_subquestions_if: true
      subquestions:
        - variable: nextcloudHostPath
          label: "Specify HostPath for Nextcloud data"
          schema:
            type: hostpath

  - variable: nextcloudDataVolume
    label: "Configure iXVolume"
    group: "Storage"
    schema:
      type: dict
      $ref:
        - "normalize/ixVolume"
      show_if: [["nextcloudDataHostPathEnabled", "=", false]]
      attrs:
        - variable: mountPath
          label: "Mount Path"
          description: "Path where the volume will be mounted inside the pod"
          schema:
            type: path
            editable: false
            default: "/var/www"
        - variable: datasetName
          label: "Dataset Name"
          schema:
            type: string
            default: "ix-nextcloud_data"
            editable: false
```
We should not be rendering `nextcloudDataVolume` variable if the first variable is true. How we know this is by the `show_if` attribute present. Please ping me for more details for this. This also included properly handling `subquestions` and `show_subquestions_if`.
4) We are not rendering GPU configuration options and this should be done as UI is supposed to be dynamically dealing with dictionaries.
```
              {
                "variable": "gpuConfiguration",
                "label": "GPU Configuration",
                "group": "Resource Reservation",
                "schema": {
                  "type": "dict",
                  "$ref": [
                    "definitions/gpuConfiguration"
                  ],
                  "attrs": [
                    {
                      "variable": "nvidia.com/gpu",
                      "label": "GPU Resource",
                      "schema": {
                        "type": "int",
                        "max": 1
                      }
                    }
                  ]
                }
```
Which makes me believe we have bits missing ?
5) Out of curiosity, can you please clarify how UI would render the following variable ?
```
  - variable: extraEnv
    label: "Environment variables for plex"
    group: "Plex Configuration"
    schema:
      type: dict
      additional_attrs: true
      attrs: []
```
Would it ignore it and not show anything ? I understand its not a good way to expose it in the first place as it's not deterministic enough but just curious on how the UI takes this.

6) I modified one argument of minio chart to the following and I get the following display
The modified version was
```
  - variable: extraArgs
    label: "Minio Extra Arguments"
    group: "Minio Configuration"
    schema:
      type: list
      default: []
      items:
        - variable: argEnv
          label: "Argument Env"
          schema:
            type: dict
            attrs:
              - variable: key
                label: "Key"
                schema:
                  type: string
              - variable: value
                label: "Value"
                schema:
                  type: string
```
Can we please correct that ?

Screenshot 2021-01-19 at 12 47 13 AM

**Comment by AlexKarpov98** [comment]
@sonicaj 
Implemented all the above 6 tasks, please check once.
Regarding No 5, currently those items are ignored and show nothing on UI.
Thank you.

**Review by sonicaj — COMMENTED** [review]
I am seeing the following issues currently:

1) After making some configuration changes for minio chart, I see in the UI
Screenshot 2021-01-20 at 5 45 05 PM

The change was

```
commit 2fb1f20e0c2699ef25406b226f9c3c4924211015
Author: Waqar Ahmed <waqarahmedjoyia@live.com>
Date:   Wed Jan 20 17:39:58 2021 +0500

    Render environment variables for minio in a more ui friendly way

diff --git a/test/minio/1.0.1/questions.yaml b/test/minio/1.0.1/questions.yaml
index f6c3121..1ec0b95 100644
--- a/test/minio/1.0.1/questions.yaml
+++ b/test/minio/1.0.1/questions.yaml
@@ -102,9 +102,24 @@ questions:
     label: "Minio image environment"
     group: "Minio Configuration"
     schema:
-      type: dict
-      additional_attrs: true
-      attrs: []
+      type: list
+      default: []
+      items:
+        - variable: environmentVariable
+          label: "Environment Variable"
+          schema:
+            type: dict
+            attrs:
+              - variable: name
+                label: "Name"
+                schema:
+                  type: string
+                  required: true
+              - variable: value
+                label: value
+                schema:
+                  type: string
+                  required: true

   - variable: service
     description: "Minio Service Configuration"
@@ -146,14 +161,14 @@ questions:
       show_if: [["minioHostPathEnabled", "=", false]]
       attrs:
         - variable: mountPath
-          label: "Mount Path"
+          label: "Minio Data Mount Path"
           description: "Path where the volume will be mounted inside the pod"
           schema:
             type: path
             editable: false
             default: "/export"
         - variable: datasetName
-          label: "Dataset Name"
+          label: "Minio Dataset Name"
           schema:
```

I think we should perhaps show the label of a dict variable which has children inside ? This would create a logical section. Like in the following screenshot, UI renders the form correctly but it does not look nice and is actually confusing because it's easy to mix fields.
Screenshot 2021-01-20 at 5 50 02 PM

In my opinion we should be showing the label of the dicts which hold these variables as well, they would group these more logically together.

2) Schema has an attribute named `editable`, these fields should be `readonly` in UI if `editable` is false, otherwise it's assumed the field is editable. Right now UI does not do that.

**Comment by AlexKarpov98** [comment]
Fixed

**Review by sonicaj — COMMENTED** [review]
Thank you for the changes, it's much much better now as we can logically group bits together now. Okay so now I see the following bits:
1) Non-editable fields are readonly now which is good but they have a cross symbol next to them which user can click on to clear the value which is not desired.
Screenshot 2021-01-21 at 1 34 58 PM

2) Secondly please have a peek at the above screenshot, do you think if it's possible to better utilize the space available ? I wonder if we can perhaps have groups sequentially and then lay out only 1 group fields at a time going over the columns. Like container images comes first in the screenshot above, then workload configuration, then minio configuration etc. Thoughts ?
Screenshot 2021-01-21 at 1 37 39 PM

**Review by sonicaj — COMMENTED** [review]
Also I see that `Minio Extra Argument` label is not being rendered, can we please look into that too ?
Screenshot 2021-01-21 at 1 40 22 PM

```
  - variable: extraArgs
    label: "Minio Extra Arguments"
    group: "Minio Configuration"
    schema:
      type: list
      default: []
      items:
        - variable: arg
          label: "Argument"
          schema:
            type: string
```

**Comment by Ornias1993** [comment]
I liked the way how the IX-Chart works like some sort of wizard.
Can't we just group things the same way for the dynamic chart ui?
(turn groups into wizard steps)

File: src/app/pages/applications/chart-releases/chart-releases.component.ts [hunk 1]
```diff
@@ -65,21 +62,45 @@ export class ChartReleasesComponent implements OnInit {
...
   }
 
-  refreshForms() {
-    this.chartReleaseForm = new ChartReleaseEditComponent(this.mdDialog,this.dialogService,this.modalService,this.appService);
-    this.plexForm = new PlexFormComponent(this.mdDialog,this.dialogService,this.modalService,this.sysGeneralService,this.appService);
-    this.nextCloudForm = new NextCloudFormComponent(this.mdDialog,this.dialogService,this.modalService,this.appService);
-    this.minioForm = new MinioFormComponent(this.mdDialog,this.dialogService,this.modalService);
+  getCatalogApps() {
+    this.appService.getAllCatalogItems().subscribe(res => {
+      res.forEach(temp => {
```
Comment (Ornias1993) [inline src/app/pages/applications/chart-releases/chart-releases.component.ts:76]: Maybe a different name than "temp"?
Something that more clearly describes what "temp"is, in this case it seems to be CatalogItems, so maybe "items" or something like it?

File: src/app/pages/applications/catalog/catalog.component.ts [hunk 2]
```diff
@@ -67,10 +62,10 @@ export class CatalogComponent implements OnInit {
 
   ngOnInit(): void {
     this.appService.getAllCatalogItems().subscribe(res => {
-      if (Object.keys(res[0].trains.charts).length > 0) {
-        for (let i in res[0].trains.charts) {  // will eventually add the charts train too
+      res.forEach(temp => {
+        for (let i in temp.trains.charts) {  
           if (i !== 'ix-chart') {
-            let item = res[0].trains.charts[i];
+            let item = temp.trains.charts[i];
```
Comment (Ornias1993) [inline src/app/pages/applications/catalog/catalog.component.ts:68]: Same as with the other use of temp

Comment (Ornias1993) [reply src/app/pages/applications/catalog/catalog.component.ts]: Solved, looks good! :)

File: src/app/pages/applications/chart-releases/chart-releases.component.ts (hunk 1 above)
Comment (Ornias1993) [reply src/app/pages/applications/chart-releases/chart-releases.component.ts]: Solved, looks good! :)

**Comment by AlexKarpov98** [comment]
Fixed

**Review by sonicaj — COMMENTED** [review]
The spacing is much better now, I have following concerns after some more testing:

1) Can we please have a way so that the user can clear the only item he/she has in the list ?
Screenshot 2021-01-21 at 11 36 29 PM

2) I am still seeing the cross button here for non-editable fields.
Screenshot 2021-01-21 at 11 37 21 PM

3) As discussed in slack, for editing a chart release, let's please use `options.extra.include_chart_schema` to retrieve the installed chart release chart details ( if you have any queries wrt this, please feel free to ping me ).

4) Installing minio catalog item, I see that we are sending a string for an integer which results in validation failing, the `service.nodePort` attribute is an integer and even the field only allows integers to be specified, however it does not send an integer to middleware.
Moving on, for an empty list it sends a value of `null` which is basically related to (1) for allowing to deselect an option so that it sends an empty list instead.
Screenshot 2021-01-22 at 3 27 05 PM
Screenshot 2021-01-22 at 3 27 12 PM

I used middleware with the same payload ( stripping out defaults ) and it created a chart release nicely
```
 ✘ ⚡ root@scale  ~  midclt call -job chart.release.create '{"catalog": "OFFICIAL", "item": "minio", "values": {"service": {"nodePort": 24122}, "secretKey": "12345678", "accessKey": "12345678"}, "release_name": "minio-ui-test"}'
Status: Chart release created
Total Progress: [########################################] 100.00%
{"name": "minio-ui-test", "info": {"first_deployed": "2021-01-22T15:30:19.957904373+05:00", "last_deployed": "2021-01-22T15:30:19.957904373+05:00", "deleted": "", "description": "Install complete", "status": "deployed", "notes": "Minio can be accessed from the following URL:\nhttp://$node_ip:24122/\n"}, "config": {"accessKey": "12345678", "environment": {}, "extraArgs": [], "image": {"pullPolicy": "IfNotPresent", "repository": "minio/minio", "tag": "RELEASE.2020-11-19T23-48-16Z"}, "ixChartContext": {}, "ixExternalInterfacesConfiguration": [], "ixExternalInterfacesConfigurationNames": [], "ixVolumes": [{"hostPath": "/mnt/evo/ix-applications/releases/minio-ui-test/volumes/ix_volumes/ix-minio", "mountPath": "/export"}], "minioDataVolume": {"datasetName": "ix-minio", "mountPath": "/export"}, "minioHostPathEnabled": false, "secretKey": "12345678", "service": {"nodePort": 24122}, "updateStrategy": "RollingUpdate"}, "version": 1, "namespace": "ix-minio-ui-test", "chart_metadata": {"name": "minio", "home": "https://min.io", "sources": ["https://github.com/minio/minio", "https://github.com/minio/charts"], "version": "1.0.0", "description": "High Performance, Kubernetes Native Object Storage", "keywords": ["storage", "object-storage", "S3"], "icon": "https://min.io/resources/img/logo/MINIO_wordmark.png", "apiVersion": "v1", "appVersion": "master", "latest_chart_version": "1.0.0"}, "id": "minio-ui-test", "catalog": "OFFICIAL", "catalog_train": "charts", "path": "/mnt/evo/ix-applications/releases/minio-ui-test", "dataset": "evo/ix-applications/releases/minio-ui-test", "status": "ACTIVE", "used_ports": [{"port": 24122, "protocol": "TCP"}], "pod_status": {"desired": 1, "available": 1}, "update_available": false, "container_images_update_available": false, "portals": {"web_portal": ["http://192.168.0.3:24122/"]}}
 ⚡ root@scale  ~ 
```

It looks to be in much better position now. I'm still experimenting more with schema to ensure we have all cases covered or at least major/common to come across cases covered.

**Review by sonicaj — COMMENTED** [review]
While playing with it, I have noticed that UI also does not render `enum` properly for integer based types, please see the following example
```
{
            "variable": "gpuConfiguration",
            "label": "GPU Configuration",
            "group": "Resource Reservation",
            "schema": {
              "type": "dict",
              "$ref": [
                "definitions/gpuConfiguration"
              ],
              "attrs": [
                {
                  "variable": "gpu.intel.com/i915",
                  "label": "GPU Resource (gpu.intel.com/i915)",
                  "description": "Please enter the number of GPUs to allocate",
                  "schema": {
                    "type": "int",
                    "max": 1,
                    "enum": [
                      {
                        "value": 0,
                        "description": "Allocate 0 GPU"
                      },
                      {
                        "value": 1,
                        "description": "Allocate 1 GPU"
                      }
                    ],
                    "default": 0
                  }
                },
                {
                  "variable": "nvidia.com/gpu",
                  "label": "GPU Resource (nvidia.com/gpu)",
                  "description": "Please enter the number of GPUs to allocate",
                  "schema": {
                    "type": "int",
                    "max": 2,
                    "enum": [
                      {
                        "value": 0,
                        "description": "Allocate 0 GPU"
                      },
                      {
                        "value": 1,
                        "description": "Allocate 1 GPU"
                      },
                      {
                        "value": 2,
                        "description": "Allocate 2 GPU"
                      }
                    ],
                    "default": 0
                  }
                }
              ]
            }
```

With the above, my expectation was that UI will show an enum like it does for Image Pull policy but it does the following:

Screenshot 2021-01-22 at 9 59 20 PM

**Comment by Ornias1993** [comment]
@sonicaj I was just going to ask about the enum, thanks for checking! :)

**Comment by AlexKarpov98** [comment]
No 1, No 2, No 4 is related with UI system but not only on Apps section issue and it will be fixed by others later.
Other issues are fixed.

**Comment by sonicaj** [comment]
@AlexKarpov98 sending a `null` on an empty list renders the user incapable of installing a chart as i shared above in my earlier comment. Can we at least please correct that in this PR ? The same is true for sending an integer on an integer field and not a string on an integer field to middleware ?

**Review by sonicaj — COMMENTED** [review]
Also can we please ensure all of these ( https://github.com/freenas/freenas/blob/a75cf27b5beb7947d6d36cad60998e9d4bad1c64/src/middlewared/middlewared/utils/__init__.py#L99 ) are supported in the UI, i tested some and not all of them work. Please let me know if there are queries/concerns, thank you

**Comment by AlexKarpov98** [comment]
Fixed this one.
https://github.com/freenas/webui/pull/5059#issuecomment-766692281

**Review by sonicaj — COMMENTED** [review]
I have the following concerns:

1) We should not be making a call with `include_chart_schema=True` unless we are editing a chart or if we do want to make it for caching purposes, we can do it in the background as this call is going to take time which means that loading apps takes a nice amount of time if you have a nice number of apps installed.

Screenshot 2021-01-27 at 12 57 51 PM

2) I am still seeing an issue where we should be sending an empty list but we are instead sending `[null]`. Please test installing minio chart as mentioned above as well.
Screenshot 2021-01-27 at 1 39 30 PM

**Review by sonicaj — COMMENTED** [review]
We hide the UI field themselves which should not be shown based on the condition specified but like in the screenshot, we still render the label for the parent and we should not render it.

Screenshot 2021-01-27 at 10 13 24 PM

```
  - variable: minioHostPathEnabled
    label: "Configure Host Path for Minio data"
    group: "Storage"
    schema:
      type: boolean
      default: false
      show_subquestions_if: true
      subquestions:
        - variable: minioHostPath
          label: "Specify HostPath for Minio data"
          schema:
            type: hostpath

  - variable: minioDataVolume
    label: "Configure iXVolume"
    group: "Storage"
    schema:
      type: dict
      $ref:
        - "normalize/ixVolume"
      show_if: [["minioHostPathEnabled", "=", false]]
      attrs:
        - variable: mountPath
          label: "Minio Data Mount Path"
          description: "Path where the volume will be mounted inside the pod"
          schema:
            type: path
            editable: false
            default: "/export"
        - variable: datasetName
          label: "Minio Dataset Name"
          schema:
            type: string
            default: "ix-minio"
            editable: false
```

We are showing `Configure iXVolume` and we should not be rendering this.

**Review by sonicaj — COMMENTED** [review]
As discussed in slack, we need to allow sending empty lists and not doing validation checks for the items the list might hold as it's possible the list is just going to be empty. Let's please allow removing the last item in the list in this case and not populate a default entry when the form is rendered for the first time.

**Comment by kmoore134** [comment]
Confirmed the issues with three dot menu and opened slide tray are resolved. Thanks @AlexKarpov98 !

**Review by sonicaj — APPROVED** [review]
Thanks!

**Review by dszidi — APPROVED** [review]
Looks ok to me. Nice work

**Comment by AlexKarpov98** [comment]
@Ornias1993 Can you please review and approve? Thank you.

**Comment by Ornias1993** [comment]
Nice work @AlexKarpov98 :)

**Pull request merged**